### ISSUE TITLE
lake does not reliably regenerate .hash files

### ISSUE THREAD

**Issue by digama0** [leanprover/lake#181]
If you delete the .hash files, lake should regenerate them, but it only seems to regenerate `olean.hash` and not `ilean.hash` and `c.hash` files, even with the `-R` flag.

**Comment by tydeu** [comment]
@digama0 The `--rehash` shorthand is `-H`, not `-R`.

**Comment by tydeu** [comment]
@digama0 More generally, when you say "delete `.hash` files" do you mean manually deleting all three of `olean.hash`, `ilean.hash`, and `c.hash`? If so, the reason it only regenerates `olean.hash` is likely because that is the only one used by downstream dependencies if the C file is not compiled (e.g.,, for a pure Lean build like mathlib). Modules are only rehashed when rebuilt so if the module is not built the hashes will not be regenerated unless they are needed by downstream dependencies.

**Comment by digama0** [comment]
> Modules are only rehashed when rebuilt

Right, this is the problem. An output having a hash should be part of what it means for the output to be up to date, so if you delete the `.c` file, it should be rebuilt, if you delete the `.c.hash` file it should be rehashed (without rerunning the C code generator), etc.

**Comment by tydeu** [comment]
@digama0 

> Currently there is no way to get out of the state where you have everything but the `.hash` files into a fully consistent state without rebuilding everything from scratch, even though it should be significantly cheaper to just hash everything.

The hash is meant as a cache, if it isn't needed there is no reason to create it.  Creating the hash externally is also not difficult it is just running the hash function on the binary/text representation of the file. I am confused as to why the absence of hash file is a problem.

**Comment by tydeu** [comment]
However, I am happy to add this if this feature is needed for some reason.

**Comment by digama0** [comment]
For context, this broke the cache, because it would download old cache files (without any hashes), recompile (producing only olean hashes) and then push cache files (sitll missing hashes), meaning that when unpacked on users' machines (containing possibly outdated/incorrect hashes) the old hashes would not be replaced, leading to an inconsistent state containing wrong hashes.

**Comment by tydeu** [comment]
@digama0 Why does cache not cleanup files before/after unpacking? The lack of this behavior seems to more generally be causing problems.

**Comment by digama0** [comment]
Because `lake exe cache` doesn't cache the whole build directory, just specific very expensive files. It's lake's job to keep the build directory in order, not `cache`'s.

**Comment by tydeu** [comment]
@digama0 Yes, but the point is that cache seems to assume what Lake's notion of in-order is, which is entirely an implementation detail on Lake's part (and can and will change rapidly between versions). If cache assumes some structure, it should enforce that for the files it cares about. That is, it should preserve it cache directory structure when unpacking/packing. For example, if is unpacking a `.hash` and is missing one then it should delete one present in the  directory and if it is packing and the file is missing it should be empty in cache (and the same should be true for other build files as well).

**Comment by digama0** [comment]
It should never be missing a .hash in the first place.

**Comment by digama0** [comment]
Hash files are a major component of that no op build performance, remember?

**Comment by tydeu** [comment]
@digama0 No they are not. In fact, they did not even change the wall-clock time in the benchmark (only task-clock). Even if they were, that would only effect the first no-op build (as they would be regenerated). To me, the cost of potential inconsistent build state is much higher than a small overhead on one build (which generally has to build some files anyway so it is an even smaller fraction for that).

**Comment by tydeu** [comment]
Honestly, downloading them like produces more overhead than producing the file itself.

**Comment by digama0** [comment]
Unpacking is done by `leantar` and is done in bulk, 4000 files in one go. It could construct hashes, but currently it's just a file store (like a .tar file as the name implies), this would require rewriting the .ltar format. It is certainly not faster than writing 8 bytes directly into a file.

**Comment by tydeu** [comment]
@digama0 Huh? I thought the cache had separate ltar's per file? If cache is running ltar on that set of files it could easily just iterate over them after they have been untarred and write the computed hash to the file, right?

**Comment by digama0** [comment]
You and I have different definitions of "easily". I don't care about how many lines of code it takes to write, I care about how long it takes to run compared to how long it should take given the constraints of the situation. The winner is very clear.

**Comment by tydeu** [comment]
@digama0

> The winner is very clear.

I disagree. Downloading and untarring a single file containing a hash is almost definitely more expensive than computing it. So,, in the worst case, computing it seems the winner. Even if the file is already local, the difference in overhead per-file would still likely be minor. Finally, computing it guarantees no inconsistent hash state in the case of cache failures or corruptions (which has happened before).

**Comment by digama0** [comment]
> Downloading and untarring a single file containing a hash is almost definitely more expensive than computing it.

* Downloading: 8 bytes (yes, really, 8 bytes). How long does that take? At a few MBs per second we're talking maybe 1 microsecond.
* Untarring: this is IO bound, so we are reading an extra 8 bytes from the file (which is already mmapped) and writing about 20 bytes into a new file (since we have to encode to decimal)

compared to:

* Read: reread the unpacked file (which is 6x larger than the input and not currently in cache), a few megabytes
* Hash: do a trivial computation over the bytes (this is an IO bound operation)
* Write: encode to decimal and write 20 bytes to a new file

There is no way you can open a megabyte large file and read it in a microsecond.

**Comment by tydeu** [comment]
@tydeu 

> * Downloading: 8 bytes (yes, really, 8 bytes). How long does that take? At a few MBs per second we're talking maybe 1 microsecond.

That is not how network download speeds work. 10% of the size is not 10% of the time. There is significant start and stop overhead for each individual file and a large part of increased speed is increased chunk size, not increased throughput (so shrinking the data below the chunk size is irrelevant). Futhermore, I am not sure why you are expecting a few MBs per second download speeds. I have really good internet for where  I live and download speeds are still often in the hundreds of KBs for me. And I know bad internet can easily be well below that (e.g., a few KBs).

**Comment by digama0** [comment]
There are no extra files, so no extra overhead for starting and stopping additional files. Again, we're making an existing file 8 bytes larger, this is a barely measurable fraction of the file size and will have negligible effect on download time.

**Comment by digama0** [comment]
per lean file, not per build output

**Comment by tydeu** [comment]
Ah
